tmt/steps/report/polarion.py

```diff
@@ -213,84 +227,115 @@ def go(self, *, logger: Optional[tmt.log.Logger] = None) -> None:
-
-        for result in self.step.plan.execute.results():
+        if logs and 'polarion-custom-logs' not in testsuites_properties:
+            testsuites_properties['polarion-custom-logs'] = logs
+
+        project_span_ids: list[str] = []
+
+        results_context = ResultsContext(self.step.plan.execute.results())
+
+        for result in results_context:
             if not result.ids or not any(result.ids.values()):
                 self.warn(
                     f"Test Case '{result.name}' is not exported to Polarion, "
                     "please run 'tmt tests export --how polarion' on it.")
                 continue
+
             work_item_id, test_project_id = find_polarion_case_ids(result.ids)
 
-            if test_project_id is None:
+            if work_item_id is None or test_project_id is None:
                 self.warn(f"Test case '{result.name}' missing or not found in Polarion.")
                 continue
 
-            assert work_item_id is not None
-            assert project_span_ids is not None
-
-            if test_project_id not in project_span_ids.attrib['value']:
-                project_span_ids.attrib['value'] += f',{test_project_id}'
+            if test_project_id not in project_span_ids:
+                project_span_ids.append(test_project_id)
 
-            test_properties = {
+            testcase_properties = {
                 'polarion-testcase-id': work_item_id,
-                'polarion-testcase-project-id': test_project_id}
+                'polarion-testcase-project-id': test_project_id,
+                }
 
-            assert testsuite is not None
-            test_case = testsuite.find(f"*[@name='{result.name}']")
-            assert test_case is not None
-            properties_elem = ElementTree.SubElement(test_case, 'properties')
-            for name, value in test_properties.items():
-                ElementTree.SubElement(properties_elem, 'property', attrib={
-                    'name': name, 'value': value})
+            result.set_properties(testcase_properties)
 
         assert self.workdir is not None
 
+        testsuites_properties.update({
+            'polarion-project-id': project_id,
+            'polarion-user-id': PolarionWorkItem._session.user_id,
+            'polarion-testrun-title': title,
+            'polarion-project-span-ids': ','.join([project_id, *project_span_ids])})
+
+        results_context.properties = testsuites_properties
+
+        xml_data = make_junit_xml(
+            phase=self,
+            flavor='polarion',
+            prettify=self.data.prettify,
+            include_output_log=self.data.include_output_log,
+            results_context=results_context,
+            )
+
         f_path = self.data.file or self.workdir / DEFAULT_NAME
-        with open(f_path, 'wb') as fw:
-            ElementTree.ElementTree(xml_tree).write(fw)
+
+        try:
+            with open(f_path, 'w') as fw:
+                fw.write(xml_data)
```

Comment:
```suggestion
            f_path.write_text(xml_data)
```
Just a nitpick, but if there is only one write operation, contextmanager and variable assignment looks redundant too me.  
(same in junit.py)

Reply:
Modified for both plugins. For junit I've added it as a part of https://github.com/teemtee/tmt/pull/3150.

Thanks.